### PR TITLE
Faraday: use FlatParamsEncoderQuery to support fetching by multiple product-versions

### PR DESCRIPTION
**Pull request by xiangge**
Query releases by product_versions with on request

JIRA: PDC-1832

### CODE REVIEW

File: pdc.gemspec [hunk 1]
```diff
@@ -34,5 +34,7 @@ Gem::Specification.new do |s|
   s.add_dependency 'faraday', '~> 0.9.2'
   s.add_dependency 'faraday_middleware', '~> 0.10.0'
   s.add_dependency 'faraday-http-cache', '~> 1.3.0'
+  # redhat sudo yum install ruby-devel libcurl-devel openssl-deve
```
Comment (sthaha) [inline pdc.gemspec]: thank you! minor typo - it is `openssl-devel`

File: spec/pdc/v1/release_spec.rb [hunk 2]
```diff
@@ -23,6 +23,11 @@
       count = release.where(active: false).count
       count.must_equal 0
     end
+
+    it 'Get release by multi product_version ' do
```
Comment (sthaha) [inline spec/pdc/v1/release_spec.rb]: Thanks for adding a test! 
Minor: To match the existing convention, it must be `it 'can get releases by multiple product_version' do ` Note `can get`

File: lib/pdc/config.rb [hunk 3]
```diff
@@ -118,7 +118,10 @@ def reset_base_connection
       faraday_config = {
         url:      api_url,
         headers:  PDC::Request.default_headers,
-        ssl:      ssl_config
+        ssl:      ssl_config,
+        request:  {
+          params_encoder: Faraday::FlatParamsEncoder
```
Comment (sthaha) [inline lib/pdc/config.rb]: None of the commit messages explain why this change is needed. Could you please amend the commit that the reason why this has been added?

Comment (xiangge) [reply lib/pdc/config.rb]: Yes, add a comment on that. This mentioned on https://github.com/lostisland/faraday#changing-how-parameters-are-serialized

File: pdc.gemspec (hunk 1 above)
Comment (xiangge) [reply pdc.gemspec]: Yes, a typo, will change that.

File: spec/pdc/v1/release_spec.rb (hunk 2 above)
Comment (xiangge) [reply spec/pdc/v1/release_spec.rb]: ok, have changed that, would please help to review it again?

**Comment by xiangge** [comment]
@sthaha thanks for your quick response! :) 
 I have finished the changes, would you please help to review it?

File: lib/pdc/config.rb [hunk 4]
```diff
@@ -118,7 +118,14 @@ def reset_base_connection
       faraday_config = {
         url:      api_url,
         headers:  PDC::Request.default_headers,
-        ssl:      ssl_config
+        ssl:      ssl_config,
+        # adding this request parameter is about changing how
```
Comment (romanofski) [inline lib/pdc/config.rb]: I like the comment, but miss the motivation of the change. Why is this changed? Was the culprit that the previous code only allowed sending single strings instead of lists?

File: spec/pdc/v1/release_spec.rb [hunk 5]
```diff
@@ -23,6 +23,11 @@
       count = release.where(active: false).count
       count.must_equal 0
     end
+
+    it 'can get release by multi product_version ' do
```
Comment (romanofski) [inline spec/pdc/v1/release_spec.rb]: Yeah honestly, with the nature of these tests it's hard to tell if it's doing the right thing. Seems alright.

File: lib/pdc/config.rb (hunk 4 above)
Comment (xiangge) [reply lib/pdc/config.rb]: Yes, the previous codes just allowed string, add this can support both list and string type of the a field.

File: spec/pdc/v1/release_spec.rb (hunk 5 above)
Comment (xiangge) [reply spec/pdc/v1/release_spec.rb]: Yes, I think basically it can test whether the url is right or not. 
Thanks for reviewing that. :)

Comment (sthaha) [inline spec/pdc/v1/release_spec.rb]: isn't this test missing the vcr files? i.e. when you run the test, doesn't it create VCR files?

Comment (xiangge) [reply spec/pdc/v1/release_spec.rb]: Yes, it creates a yml file.
Do you mean this can_get_release_by_multi_product_version_.yml? It is automatically generated.
And one place need to change, shouldn't add the blank at the end of the name. After the change the name should be can_get_release_by_multi_product_version.yml, please help to review.

Comment (sthaha) [reply spec/pdc/v1/release_spec.rb]: @xiangge those files need to be checked in just like the others, otherwise the test results can vary depending on the data in the server which isn't a behaviour we want.

**Comment by sthaha** [comment]
@xiangge thanks for adding those VCR files. I do agree with @simozhan that the commits can be better arranged/squashed to make it easy to follow. However I don't see this as a blocker as PR will create a merge-commit and is easy to revert the entire feature.

**Comment by xiangge** [comment]
@sthaha Yes, also agree with @simozhan's recommendation and thank you for your approval.
I didn't commit codes with ammend every time, because maybe you might have issue during the pull.
I've been thinking to make these  commits in one after you approve that, but seems I am slower than you, you already merged, sorry about that  :).

**Comment by sthaha** [comment]
> I didn't commit codes with ammend every time, because maybe you might have issue during the pull.

@xiangge no, it shouldn't create any issue when you amend the commit. you just have to make sure that you do a force push to **your branch**

e.g.

```shell
git commit -m 'fixup: commit ~1'
git commit -m 'fixup: foobar ~2'

git rebase --interactive upstream/master

# squash the commits
```

```shell
git push -f
```
would update the PR

**Comment by xiangge** [comment]
> you just have to make sure that you do a force push to your branch
Yes, this is what I usually do.

**Comment by sthaha** [comment]
@xiangge I just noticed that you are using a branch on this repo for feature development instead of forking the repo. For the next PR, could you please fork the repo and submit a PR?
